fix(onload): Add onload to window correlated with element

brdlyptrs opened the pull request:
Looks for the window object correlated to the element being used for loading the script.

zoryana94 reviewed (commented):
It requires additional testing
It doesn't seem to work with how I set it up locally inside the iframe

zoryana94 commented:
@brdlyptrs it looks like scriptContainerWindow.hcaptcha.render... should be called instead

So, I received `hcaptcha is undefined`
To fix it, I updated `hcaptcha` to `scriptContainerWindow.hcaptcha` locally
It's working, but I receive the "invalid-data" error-code
it seems we require some other changes related to window

Also it doesn't seem to be the best solution (just for you to be aware of the issue and what partially fixes it)

brdlyptrs commented:
Hi @zoryana94, good catch on this! 

I appreciate the example code as well, I did some further testing and will have a solution pushed up. The code you provided does not lock the hCaptcha instance to the window which is needed here, given a user could have multiple hCaptcha instances in various iFrames.

brdlyptrs commented:
After further testing, I found some edge cases as well. I've now updated the code to handle the following
- Utilize the hCaptcha API loaded relative to the document the script was appended
- Add the ability to track each hCaptcha script being loaded in order to handle multiple documents that may instantiate hCaptcha at the same time

I've also updated the demo examples to include loading via iFrame, see screenshot.

![image](https://user-images.githubusercontent.com/40441278/230683310-8c0650f7-0560-4d4a-81c6-5376f7148b1d.png)

faris-imi commented on src/index.js:
```diff
@@ -5,52 +5,68 @@ const SCRIPT_ID = 'hcaptcha-api-script-id';
 
-  script.onerror = (event) => rejectFn('script-error');
+    const query = generateQuery(params);
+    script.src += query !== ""? `&${query}` : "";
```
missing whitespace `query !== "" ? ...`

brdlyptrs replied:
@faris-imi I can follow up with this update in a separate PR

faris-imi commented on demo/app/examples/FrameExample.jsx:
```diff
@@ -0,0 +1,70 @@
+export function FrameExample({ document }) {
+  const captchaRef = useRef();
+
+  const executeCaptcha = async () => {
```
should we remove these unused functions ?

zoryana94 replied:
+ useEffect, useState

brdlyptrs replied:
addressed here [3f12a29](https://github.com/hCaptcha/react-hcaptcha/pull/185/commits/3f12a29d517a2821b4ef5bad72bd075b544fe5c9)

faris-imi commented on demo/app/examples/FrameExample.jsx:
i noticed cursor is default instead of pointer when hovering over frame challenge as it's not wrapped in div, is this expected behavior for frame example

zoryana94 replied:
I see the pointer cursor. It changes to the initial one after passing the challenge.

brdlyptrs replied:
Going to resolve this for now, its outside the scope for this.

zoryana94 commented on demo/app/App.jsx:
```diff
@@ -0,0 +1,51 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { createPortal } from 'react-dom';
```
please remove not used `useRef` and `createPortal`

joao-imi replied:
same, they're not being used in this file at all

brdlyptrs replied:
addressed here [3f12a29](https://github.com/hCaptcha/react-hcaptcha/pull/185/commits/3f12a29d517a2821b4ef5bad72bd075b544fe5c9)

zoryana94 commented on src/index.js:
```diff
@@ -245,6 +272,7 @@ class HCaptcha extends React.Component {
     handleExpire () {
       const { onExpire } = this.props;
       const { captchaId } = this.state;
+      const hcaptcha = this._hcaptcha;
```
why not to use `this._hcaptcha` directly?
`this._hcaptcha.reset(captchaId);`

joao-imi replied:
It's not necessary to remove the word `this` from variable names, but it depends on the coding conventions and style guide that we want to follow.

I like removing the `this` from the `hcaptcha` variable name and so on, it makes the code a bit more readable - but I think this a personal preference only.

Since this is a "public react library", it should be clean and easy for developers to interprete.

brdlyptrs replied:
Less legacy code to update, was my thought

zoryana94 commented on src/index.js:
```diff
@@ -5,52 +5,68 @@ const SCRIPT_ID = 'hcaptcha-api-script-id';
       super(props);
 
+      const parent = this.props.scriptLocation || document.head;
+      const doc = parent.ownerDocument || document;
```
it'd be good to create separate `getDocument()` and `getWindow()`
as this logic is repeated in 3 places right now: `mountCaptchaScript`, `handleOnLoad` and here

brdlyptrs replied:
address here [3f12a29](https://github.com/hCaptcha/react-hcaptcha/pull/185/commits/3f12a29d517a2821b4ef5bad72bd075b544fe5c9)

zoryana94 commented on src/index.js:
```diff
@@ -5,52 +5,68 @@ const SCRIPT_ID = 'hcaptcha-api-script-id';
 
+      const parent = this.props.scriptLocation || document.head;
+      const doc = parent.ownerDocument || document;
+      const win = doc.defaultView || doc.parentWindow;
```
how about renaming `win` to `hcaptchaWindow`?

brdlyptrs replied:
I wouldn't consider hCaptcha's window, its still the window maybe frameWindow is better here. This should be addressed here [3f12a29](https://github.com/hCaptcha/react-hcaptcha/pull/185/commits/3f12a29d517a2821b4ef5bad72bd075b544fe5c9)

zoryana94 replied:
It's not always the frame though
this naming is a bit confusing to me

maybe, `hcaptchaApiWindow` or `apiWindow` would suit better

brdlyptrs replied:
Thats why I do 

```
const frame = getFrame();
// frame.document
// frame.window
```

joao-imi commented:
@brad, besides those minor comments suggestions, all the features are working great on my tests!

I also just noticed that the `iframe` height is being a bit cut in the **Frame example** - which it can be a little bit weird for the demo.
We may want to set a minimum height size for the iframe demo example or stretch it automatically.

Check video below, please:

https://user-images.githubusercontent.com/128425298/230934845-4627aa30-a872-49ab-9dcc-3d82215046dd.mov

zoryana94 reviewed (dismissed):
Working well 👍

brdlyptrs commented on src/index.js:
```diff
@@ -56,16 +54,15 @@ class HCaptcha extends React.Component {
       const parent = this.props.scriptLocation || document.head;
-      const doc = parent.ownerDocument || document;
-      const win = doc.defaultView || doc.parentWindow;
+      const frame = getFrame(parent);
```
Added an update to make this a function.

zoryana94 reviewed (approved):
Re-tested 👍
